File: bindings/jvm/src/main/java/com/goplasmatic/datalogic/TracedRun.java
/* SPDX-License-Identifier: Apache-2.0 */
package com.goplasmatic.datalogic;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.MissingNode;

/**
 * Result of a traced evaluation. Mirrors the cross-binding wire JSON
 * shape: {@code {result, expression_tree, steps, error?, structured_error?}}.
 */
public final class TracedRun {
    private final JsonNode result;
    private final JsonNode expressionTree;
    private final JsonNode steps;
    private final String error;
    private final JsonNode structuredError;

    TracedRun(JsonNode result, JsonNode expressionTree, JsonNode steps,
              String error, JsonNode structuredError) {
        this.result = result == null ? MissingNode.getInstance() : result;
        this.expressionTree = expressionTree == null ? MissingNode.getInstance() : expressionTree;
        this.steps = steps == null ? MissingNode.getInstance() : steps;
        this.error = error;
        this.structuredError = structuredError == null ? MissingNode.getInstance() : structuredError;
    }

    /** Evaluation result, or a Null node if the run errored. */
    public JsonNode result() { return result; }

    /** Compile-time expression tree for flow-diagram rendering. */
    public JsonNode expressionTree() { return expressionTree; }

    /** Per-node execution log captured during the run (always a JSON array). */
    public JsonNode steps() { return steps; }

    /** Engine error message, or {@code null} on success. */
    public String error() { return error; }

    /** Structured error object, or a Missing node on success. */
    public JsonNode structuredError() { return structuredError; }

    /** Whether the run succeeded. */
    public boolean isSuccess() { return error == null; }
}

File: bindings/php/src/TracedRun.php
<?php

declare(strict_types=1);

/* SPDX-License-Identifier: Apache-2.0 */

namespace Goplasmatic\Datalogic;

/**
 * Result of a traced evaluation. Mirrors the cross-binding wire JSON
 * shape: `{result, expression_tree, steps, error?, structured_error?}`.
 */
final class TracedRun
{
    /**
     * @param mixed         $result          parsed result (any JSON value), or null on failure
     * @param mixed         $expressionTree  compile-time expression tree
     * @param list<mixed>   $steps           per-node execution log
     * @param string|null   $error           engine error message, or null on success
     * @param mixed         $structuredError structured error object, or null on success
     */
    public function __construct(
        public readonly mixed $result,
        public readonly mixed $expressionTree,
        public readonly array $steps,
        public readonly ?string $error,
        public readonly mixed $structuredError,
    ) {}

    public function isSuccess(): bool
    {
        return $this->error === null;
    }
}
